### ISSUE TITLE
Proposal: Apply type inference to redirecting factory constructors

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#366]
Cf. dart-lang/sdk#37037 where this issue came up. Consider the following situation:

```dart

class A<X extends num> {
  factory A() = B;
}

class B<X extends num> implements A<X> {}
```

In `A`, the factory constructor redirects to `B`. Given that `B` is a raw type it could mean `B<num>`, and in that case we'd have a compile-time error because the factory attempts to return an object whose type isn't a subtype of the enclosing class (`B<num>` is not a subtype of `A<X>`).

That is also how the program is treated by the common front end.

However, the analyzer does actually accept the above program, and a similar program shows that it uses type inference to make that redirection mean `B<X>` rather than `B` (cf. [this comment](https://github.com/dart-lang/sdk/issues/37037#issuecomment-494416408)).

The common front end does _not_ use type inference on the indirection, but treats it as a raw type and reports the corresponding errors.

This issue is a proposal for resolving that discrepancy in favor of using type inference.

Given that this is (very likely to be) the rule that the analyzer already implements, it would only be a breaking change for programs that were already rejected by the analyzer, and we do not normally consider that to be a breaking change.

@lrhn, @leafpetersen, @munificent, do you support this proposal?

**Comment by leafpetersen** [comment]
Yes, I think this is just a known bug in the CFE - see here: https://github.com/dart-lang/sdk/issues/34714 .

**Comment by eernstg** [comment]
.. or a bug in the analyzer, until we make a decision. ;)

**Comment by leafpetersen** [comment]
Yes, the analyzer is correct.

**Comment by lrhn** [comment]
Agree, let's do the inference here. 
The analyzer is now officially correct, the spec and CFE are lacking.

**Comment by leafpetersen** [comment]
This still seems to be an open issue.

**Comment by eernstg** [comment]
I think there is support for the approach where the raw redirectee receives actual type arguments by inference such that the resulting object has a subtype of the enclosing class: @leafpetersen [thinks](https://github.com/dart-lang/language/issues/366#issuecomment-494478880) this is already what the language should do, and @lrhn [supports](https://github.com/dart-lang/language/issues/366#issuecomment-494789046) using inference.

So I'd like to conclude this such that we can get it implemented: The omitted type arguments will be inferred.

@leafpetersen, @munificent, @lrhn, agreed?

**Comment by lrhn** [comment]
Agree

**Comment by leafpetersen** [comment]
Agreed.  There seems to be additional inconsistency as well.  Given the following:

```dart
class A<T> {
  A();
  factory A.foo(int x) = B.foo<T>;
  factory A.bar(int x) = B.bar;
}

class B<T> extends A<T> {
  B.foo(int x) {}
  B.bar(int x) {}
}
void main() {
}
```

The CFE accepts the definition of `A.foo` but not `A.bar` whereas the analyzer rejects `A.foo` and not `A.bar`.

Should `A.foo` be valid as written?

**Comment by eernstg** [comment]
I believe it would be unhelpful to allow `B.foo<T>` in this context, because it creates a conflict with generic constructors (a feature that we've had on the wish list for a long time).

For example, in instance creations like `B.foo<int>()` it would introduce an ambiguity: `B.foo<int>()` could be an invocation of a generic constructor, or it might be an invocation of the constructor `B.foo` which will yield an instance of `B<int>` (as implied in the example), and if `foo` were a generic constructor then we would presumably infer its type arguments in the latter case. Similarly, `B.foo<int>` might be supported as the syntax for a generic function instantiation (it would tear off a generic constructor specialized to have type argument list `<int>`, and it would use inference to select actual type arguments to `B`); but if we allow the class type arguments to be provided after the name `foo`, `B.foo<int>` would then be ambiguous: it might as well denote a constructor tear-off yielding an instance of `B<int>` (which would then be a generic function iff `foo` were generic).

So I don't think we should modify the analyzer to accept the class type argument list to be passed after the name of a named constructor, I think we should modify the CFE to reject this form.

Note that the CFE _does_ emit an error when the non-standard placement of the actual class type argument list is used in an instance creation: `new Foo.bar<int>()` makes the CFE respond with 'A constructor invocation can't have type arguments on the constructor name' [here](https://github.com/dart-lang/sdk/blob/33056425aff7c3bfff29410b4490ea2eec6c3e34/tests/language_2/constructor/reference_test.dart#L28), so the fact that it is allowed in the body of a redirecting factory constructor is clearly an exception.


Just to elucidate other aspects of the example, I changed `B.foo<int>` to `B<int>.foo`. The analyzer then does not report any errors (which is consistent with the assumption that the analyzer already uses inference) and no further issues arise.

After the change, the CFE still does not report an error in line 3, but proceeds to report an error in line 4, which fits the assumption that the CFE uses instantiation to bound.

So we may still **conclude** the following: A raw redirectee in a redirecting factory constructor will receive the omitted actual type arguments from type inference (not instantiation to bound).

**Comment by eernstg** [comment]
Actually, it looks unlikely that we'd want to keep `B.foo<T>` a CFE non-error in the example [here](https://github.com/dart-lang/language/issues/366#issuecomment-625404618), in particular because the CFE already reports an error for instance creations like `new B.foo<int>()`. So I created https://github.com/dart-lang/sdk/issues/41842 to get the error implemented.

**Comment by lrhn** [comment]
No, it should not be valid as written. 
It's not a grammar error, but only because `B.foo<T>` could be a prefix `B` and a class name `foo`.
It is a compile-time error because `B.foo` does not denote a *class*, and you can't put class type parameters on a named constructor name, they have to go on the class.

So, to be valid, it should be written as:
```dart
factory A.foo(int x) = B<T>.foo;
```

**Comment by eernstg** [comment]
Closing: This inference step was implemented in 3b93b85aeff4d0beb8b262044a12dff1af56df16, cf. dart-lang/sdk#34714, and it was implemented in the analyzer already before this issue was created.
